test/jdk/java/net/httpclient/CancelledResponse2.java

```diff
@@ -0,0 +1,258 @@
+import org.testng.annotations.AfterTest;
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.net.http.HttpClient;
+import java.net.http.HttpClient.Version;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandler;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Flow;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static java.lang.System.out;
+import static java.net.http.HttpClient.Version.*;
+import static jdk.httpclient.test.lib.common.HttpServerAdapters.*;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+/*
+ * @test
+ * @library /test/lib /test/jdk/java/net/httpclient/lib
+ * @build jdk.test.lib.net.SimpleSSLContext
+ * @run testng/othervm -Djdk.internal.httpclient.debug=true CancelledResponse2
+ */
+
+public class CancelledResponse2 {
+
+
+    HttpTestServer h2TestServer;
+    URI h2TestServerURI;
+    private SSLContext sslContext;
+    private static final Random random = RandomFactory.getRandom();
+    private static final int MAX_CLIENT_DELAY = 160;
+
+    @DataProvider(name = "versions")
+    public Object[][] positive() {
+        return new Object[][]{
+                { HTTP_2, h2TestServerURI },
+        };
+    }
+
+    private static void delay() {
+        int delay = random.nextInt(MAX_CLIENT_DELAY);
+        try {
+            System.out.println("client delay: " + delay);
+            Thread.sleep(delay);
+        } catch (InterruptedException x) {
+            out.println("Unexpected exception: " + x);
+        }
+    }
+
+    @Test(dataProvider = "versions")
+    public void test(Version version, URI uri) throws Exception {
+        for (int i = 0; i < 5; i++) {
+            HttpClient httpClient = HttpClient.newBuilder().sslContext(sslContext).version(version).build();
+            HttpRequest httpRequest = HttpRequest.newBuilder(uri)
+                    .version(version)
+                    .GET()
+                    .build();
+            AtomicBoolean cancelled = new AtomicBoolean();
+            BodyHandler<String> bh = ofString(response, cancelled);
+            CompletableFuture<HttpResponse<String>> cf = httpClient.sendAsync(httpRequest, bh);
+            try {
+                cf.get();
+            } catch (Exception e) {
+                e.printStackTrace();
+                assertTrue(e.getCause() instanceof IOException, "HTTP/2 should cancel with an IOException when the Subscription is cancelled.");
```

Comment:
I believe we want to check that the exception we get at the cause is the expected CancelException, or at least that it can be found somewhere while following the cause chain. Otherwise looks good!

Reply:
There is a check that the cancelling BodyHandler sets the cancelled field which may be sufficient

Reply:
Could also just check that the returned a call to `getMessage()` returns a string that contains cancelled but its not a very nice solution

Reply:
Update: can verify the subscription is properly cancelled by adding a reference to the BodySubscriber used to the CancellingHandler (client) in the test. When the test is done, all thats needed is a call to `SubscriptionWrapper.cancelled()` to verify that the connection ended correctly.